tiledb/sm/cpp_api/subarray.h

```diff
@@ -0,0 +1,695 @@
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * This file declares the C++ API for the TileDB Subarray object.
+ */
+
+#ifndef TILEDB_CPP_API_SUBARRAY_H
+#define TILEDB_CPP_API_SUBARRAY_H
+
+#include "array.h"
+#include "array_schema.h"
+#include "context.h"
+#include "core_interface.h"
+#include "deleter.h"
+#include "exception.h"
+#include "tiledb.h"
+#include "type.h"
+#include "utils.h"
+
+#include <algorithm>
+#include <cassert>
+#include <functional>
+#include <iterator>
+#include <memory>
+#include <set>
+#include <string>
+#include <type_traits>
+#include <unordered_map>
+#include <vector>
+
+namespace tiledb {
+
+class Query;
+
+/**
+ * Construct and support manipulation of a possibly multiple-range subarray for
+ * optional use with Query object operations.
+ *
+ * @details
+ * See examples for more usage details.
+ *
+ * **Example:**
+ * @code{.cpp}
+ * // Open the array for writing
+ * tiledb::Context ctx;
+ * tiledb::Array array(ctx, "my_dense_array", TILEDB_WRITE);
+ * Query query(ctx, array);
+ * std::vector a1_data = {1, 2, 3};
+ * query.set_buffer("a1", a1_data);
+ * tiledb::Subarray subarray(ctx, array);
+ * subarray.set_layout(TILEDB_GLOBAL_ORDER);
+ * std::vector<int32_t> subarray_indices = {1, 2};
+ * subarray.add_range(0, subarray_indices[0], subarray_indices[1]);
+ * query.set_subarray(subarray);
+ * query.submit();
+ * query.finalize();
+ * array.close();
+ * @endcode
+ */
+class Subarray {
+ public:
+  Subarray(
+      const tiledb::Context& ctx,
+      const tiledb::Array& array,
+      bool coalesce_ranges = true)
+      : ctx_(ctx)
+      , array_(array)
+      , schema_(array.schema()) {
+    tiledb_subarray_t* capi_subarray;
+    ctx.handle_error(tiledb_subarray_alloc(
+        ctx.ptr().get(), array.ptr().get(), &capi_subarray));
+    tiledb_subarray_set_coalesce_ranges(
+        ctx.ptr().get(), capi_subarray, coalesce_ranges);
+    subarray_ = std::shared_ptr<tiledb_subarray_t>(capi_subarray, deleter_);
+  }
+
+  Subarray(const tiledb::Query& query);  // defined in cppapi Query.h
```

Comment:
The caller can construct a `Subarray` with `query.ctx()` and `query.array()`. Remove this constructor and its implementation in `Query.h`.

Reply:
Construction with query.ctx() and query.array() does not achieve same functionality, as that will only have default subarray state where as this constructor produced Subarray with the query's current subarray state.
Have however removed constructor and instead provided methods so same end result can be achieved though requiring client to make an additional method call.